### PR TITLE
HPCC-17214 Refactor code to determine where to evaluate cses

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by ghalliday** [comment]
@shamser 
This was tested by ensuring that all the regression suites generated identical code.  That gives me a high degree of confidence that the change is correct.  (Any slight discrepancy was highlighted in many of the tests changing).

This has little effect on many queries, but on some it is dramatic.  (E.g., see jake40good where it reduces the compile time from ~320s to < 50s).

There are a large number of comments preceded with REVISIT.  These have deliberately been left in, and will be addressed by HPCC-17215.  This PR is purely refactoring the algorithm with identical results.

File: ecl/hqlcpp/hqlcse.ipp [hunk 1]
```diff
@@ -95,102 +93,156 @@ public:
...
+    CseScopeInfo * inheritedAliases = nullptr;// Which expression aliases should be inherited from - saves appending to childAliases for common case.
+    ICopyArrayOf<CseScopeInfo> parentAliases;// List of all aliases which this alias is used by.  (Not set for non-aliases)
+    unsigned seq = 0;                        // A unique sequence number used for tracing and finding a common parent
+    unsigned minDepth = (unsigned)-1;        // REVIST: Not sure this actually works!  Currently unused.  May need a shortest path pointer instead.
+    bool isUnconditional = false;            // Is this expression unconditional = not currently used.
+    bool isAlias = false;                    // Is this an alias?
+
+private:
+    //The following are calculated after walking the graph
+    bool calcedCommonLocation = false;       // Has the derived location information been calcualted for this node?
```
Comment (shamser) [inline ecl/hqlcpp/hqlcse.ipp]: @ghalliday spelling "calcualted"

File: ecl/hqlcpp/hqlcse.ipp [hunk 2]
```diff
@@ -95,102 +93,156 @@ public:
...
+    bool moved = false;                      // Should this expression be evalauted in a different location in the graph?
```
Comment (shamser) [inline ecl/hqlcpp/hqlcse.ipp]: @ghalliday spelling "evalauted"

File: ecl/hqlcpp/hqlcerrors.hpp [hunk 3]
```diff
@@ -224,6 +224,7 @@
 #define HQLERR_BodyNotAllowedWithInline         4204
 #define HQLERR_DatasetPassedToRowArg            4205
 #define HQLERR_HashStoredDuplication            4207
+#define HQLERR_RankOnNonList                    4208
```
Comment (shamser) [inline ecl/hqlcpp/hqlcerrors.hpp]: @ghalliday No associated HQLERR_RankOnNonList_Text.  Also, this is unused for now.

File: ecl/hqlcpp/hqlcse.ipp [hunk 4]
```diff
@@ -95,102 +93,156 @@ public:
...
-    Owned<CSEentry>     cseUse;
-    CIArrayOf<CSEentry> cseDefine;
+    HqlExprArray aliasesToDefine;            // Which aliases need to be evaluated when this expression is evaluated
+    CseScopeInfo * firstParent = nullptr;    // use a pointer for the first match to avoid reallocation for unshared items
+    PointerArrayOf<CseScopeInfo> parents;    // list of all parent expressions - other than firstParent
+    ICopyArrayOf<CseScopeInfo> childAliases; // List of child aliases - interpret in conjunction with inheritedAliases
+    CseScopeInfo * inheritedAliases = nullptr;// Which expression aliases should be inherited from - saves appending to childAliases for common case.
+    ICopyArrayOf<CseScopeInfo> parentAliases;// List of all aliases which this alias is used by.  (Not set for non-aliases)
+    unsigned seq = 0;                        // A unique sequence number used for tracing and finding a common parent
+    unsigned minDepth = (unsigned)-1;        // REVIST: Not sure this actually works!  Currently unused.  May need a shortest path pointer instead.
```
Comment (shamser) [inline ecl/hqlcpp/hqlcse.ipp]: @ghalliday Minor: could use the UINT_MAX constant

**Review by shamser — COMMENTED** [review]
@ghalliday A few minor comments.

File: ecl/hqlcpp/hqlcpp.cpp [hunk 5]
```diff
@@ -9466,6 +9466,12 @@ void checkRankRange(IHqlExpression * index, IHqlExpression * list)
         getStoredDescription(s, sequence, name, true);
         throwError1(HQLERR_RankOnStored, s.str());
     }
+    if (list->getOperator() != no_list)
+    {
+        StringBuffer s;
+        getExprECL(list, s);
+        throwError1(HQLERR_RankOnStored, s.str());
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlcpp.cpp]: I will rename it to RankOnNonList and delete the extra #define

**Comment by ghalliday** [comment]
@shamser PR updated and changes squashed since they are all trivial.

**Review by shamser — APPROVED** [review]
@richardkchapman Code reviewed ok.

**Pull request merged**